### PR TITLE
Refactor rules loading result

### PR DESCRIPTION
**Pull request by mstemm**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> If contributing rules or changes to rules, please make sure to also uncomment one of the following line:

> /kind rule-update

> /kind rule-create

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area rules

/area tests

> /area proposals

**What this PR does / why we need it**:

When loading rules, return a structure with the load result and lists of errors/warnings + file locations.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
update: when loading rules, return a structure with the load result and lists of errors/warnings + file locations.
```

### CODE REVIEW

**Comment by mstemm** [comment]
One follow-on change I'd like to make is to extend the context struct to point inside a condition string/output string. I think that might require some libs changes such as annotating ASTs with yaml file positions. I see that the parser object has a position but I think for multi-line strings the yaml parsing changes the position. Let me know if I'm wrong on that and I can work on merging the positions.

**Comment by mstemm** [comment]
I cherry-picked the commit from https://github.com/falcosecurity/falco/pull/2096 and included it in this PR temporarily, just so I could get the build/test working. Once 2096 is merged, I'll rebase and drop the cherry-picked commit.

**Comment by mstemm** [comment]
Here are some example outputs that use the result structure and print the contents in various ways:

Validating some files, no errors:

```
$ ./userspace/falco/falco -V /etc/falco/falco_rules.yaml -V /etc/falco/falco_rules.local.yaml
Thu Jun 30 08:32:49 2022: Falco version 0.0.0
Thu Jun 30 08:32:49 2022: Falco initialized with configuration file /etc/falco/falco.yaml
Thu Jun 30 08:32:49 2022: Validating rules file(s):
Thu Jun 30 08:32:49 2022:    /etc/falco/falco_rules.yaml
Thu Jun 30 08:32:49 2022:    /etc/falco/falco_rules.local.yaml
/etc/falco/falco_rules.yaml: Ok
/etc/falco/falco_rules.local.yaml: Ok
```

Validating some files, some errors + warnings. This version of the output is terse, which matches old behavior:

```
$ ./userspace/falco/falco -V /etc/falco/falco_rules.yaml -V /etc/falco/falco_rules.local.yaml
Thu Jun 30 08:35:19 2022: Falco version 0.0.0
Thu Jun 30 08:35:19 2022: Falco initialized with configuration file /etc/falco/falco.yaml
Thu Jun 30 08:35:19 2022: Validating rules file(s):
Thu Jun 30 08:35:19 2022:    /etc/falco/falco_rules.yaml
Thu Jun 30 08:35:19 2022:    /etc/falco/falco_rules.local.yaml
Error: /etc/falco/falco_rules.yaml: Ok
/etc/falco/falco_rules.local.yaml: Invalid
```

If you add `-v` you get full details on the errors + warnings:

```
$ ./userspace/falco/falco -V /etc/falco/falco_rules.yaml -V /etc/falco/falco_rules.local.yaml -v
Thu Jun 30 08:38:48 2022: Falco version 0.0.0
Thu Jun 30 08:38:48 2022: Falco initialized with configuration file /etc/falco/falco.yaml
Thu Jun 30 08:38:48 2022: Validating rules file(s):
Thu Jun 30 08:38:48 2022:    /etc/falco/falco_rules.yaml
Thu Jun 30 08:38:48 2022:    /etc/falco/falco_rules.local.yaml
/etc/falco/falco_rules.local.yaml: Invalid
1 Errors:
In rule 'My Invalid Rule': (/etc/falco/falco_rules.local.yaml:37:2)
    exception 'invalid exception': (/etc/falco/falco_rules.local.yaml:43:6)
------
    - name: invalid exception
      ^
------
FE_LOAD_ERR_VALIDATE (Error validating rule/macro/list/exception objects): 'not-a-field' is not a supported filter field

1 Warnings:
In rule 'My Warning Rule': (/etc/falco/falco_rules.local.yaml:31:2)
------
- rule: My Warning Rule
  ^
------
FE_LOAD_NO_EVTTYPE (Condition has no event-type restriction): Rule matches too many evt.type values. This has a significant performance penalty.

Error: /etc/falco/falco_rules.yaml: Ok
/etc/falco/falco_rules.local.yaml: Invalid
```

And if you add `-o json_output=True` you get a json version of the errors + warnings instead. The result is printed to stdout, so if you redirect stderr you get *only* the validation result. This matches older falco behavior:

```
$ ./userspace/falco/falco -V /etc/falco/falco_rules.yaml -V /etc/falco/falco_rules.local.yaml -v -o json_output=True 2>/dev/null | jq "."
{
  "falco_load_results": [
    {
      "errors": [],
      "name": "/etc/falco/falco_rules.yaml",
      "successful": true,
      "warnings": []
    },
    {
      "errors": [
        {
          "code": "FE_LOAD_ERR_VALIDATE",
          "codedesc": "Error validating rule/macro/list/exception objects",
          "context": {
            "locations": [
              {
                "item_name": "My Invalid Rule",
                "item_type": "rule",
                "position": {
                  "column": 2,
                  "filename": "/etc/falco/falco_rules.local.yaml",
                  "line": 37,
                  "offset": 1246
                }
              },
              {
                "item_name": "invalid exception",
                "item_type": "exception",
                "position": {
                  "column": 6,
                  "filename": "/etc/falco/falco_rules.local.yaml",
                  "line": 43,
                  "offset": 1374
                }
              }
            ],
            "snippet": "    - name: invalid exception\n      ^\n"
          },
          "message": "'not-a-field' is not a supported filter field"
        }
      ],
      "name": "/etc/falco/falco_rules.local.yaml",
      "successful": false,
      "warnings": [
        {
          "code": "FE_LOAD_NO_EVTTYPE",
          "codedesc": "A rule condition matches too many evt.type values. This has a significant performance penalty. Make the condition more specific by adding an evt.type field or further restricting the number of evt.type values in the condition.",
          "context": {
            "locations": [
              {
                "item_name": "My Warning Rule",
                "item_type": "rule",
                "position": {
                  "column": 2,
                  "filename": "/etc/falco/falco_rules.local.yaml",
                  "line": 31,
                  "offset": 1139
                }
              }
            ],
            "snippet": "- rule: My Warning Rule\n  ^\n"
          },
          "message": "Rule matches too many evt.type values. This has a significant performance penalty."
        }
      ]
    }
  ]
}
```

File: test/falco_tests.yaml
Comment (jasondellaluce) [inline test/falco_tests.yaml:25]: For context, this regression comes from way before the C++ porting (removal of `sinsp_rule_utils.lua`): https://github.com/falcosecurity/falco/pull/1715

File: test/falco_tests.yaml
Comment (jasondellaluce) [inline test/falco_tests.yaml:465]: This probably connects to my other comment: here the warning is not meaningful. We should have failed at the first error.

File: test/falco_tests.yaml
Comment (jasondellaluce) [inline test/falco_tests.yaml:1028]: Why is this removed? Same goes for `macro_append_failure` and `rule_append_failure`.

File: test/falco_tests.yaml
Comment (jasondellaluce) [inline test/falco_tests.yaml:34]: Are these all of type `LOAD_NO_EVTTYPE`? Or will we need to change them once we bring back the other kinds of warnings we had in `sinsp_rule_utils.lua`?

File: test/falco_test.py
Comment (jasondellaluce) [inline test/falco_test.py:389]: So basically `message` and `message_contains` seem to be mandatory, right? Should we make it optional? Should we document this somewhere?

**Review by jasondellaluce — COMMENTED** [review]
Thanks for this effort Mark! I tried my best to review this, left you some comments.

On top of that, I think we also need to change the documentation here: https://github.com/falcosecurity/falco/blob/35db0b4a24344bc6c24022555f7d8531ad925136/falco.yaml#L77

`json_output` now does more than just formatting rule outputs, so I think we should document it properly. Alternatively, we should further separate responsibilities and have a separate config field like `json_load_result`.

**Comment by mstemm** [comment]
I created a companion docs PR https://github.com/falcosecurity/falco-website/pull/657 that updates the docs for json_output to reflect that it controls both alerts and rule loading/validation results. I think a single option is preferable so I updated the docs rather than split into two separate options.

(I also updated the comments in falco.yaml)

File: userspace/engine/falco_load_result.h
Comment (mstemm) [reply userspace/engine/falco_load_result.h:26]: Sure, that's fine. I guess the values are still namespaced within the class so there isn't any chance of conflicts.

File: userspace/engine/falco_load_result.h
Comment (mstemm) [reply userspace/engine/falco_load_result.h:23]: Sure, I think we should put it in a falco namespace though as load_result by itself is generic. That brings up the question of whether we should add a namespace to falco_engine as well? The engine doesn't have any namespace at the moment, which is why I didn't use one for falco_load_result.

I don't want to get carried away with changes, so how about this--in this PR I simply add a falco namespace around load_result, and once this is merged we add a top level namespace to falco_engine? Maybe with an additional rename of the falco_engine class, we can bikeshed about it in the follow-on PR.

File: userspace/engine/falco_load_result.h
Comment (mstemm) [reply userspace/engine/falco_load_result.h:45]: Yep, removed in both.

File: userspace/engine/rule_loader.h
Comment (mstemm) [reply userspace/engine/rule_loader.h:91]: Sure, fixed.

File: userspace/engine/rule_loader.h
Comment (mstemm) [reply userspace/engine/rule_loader.h:72]: I seem to be the only one who likes lists, I get this question whenever I use them :) I always choose lists unless I really want the O(1) random access. But everyone else likes vectors so I can change it.

File: userspace/engine/rule_loader.h
Comment (mstemm) [reply userspace/engine/rule_loader.h:62]: Yeah, I went back and forth on it, thinking that snippet by itself would be too generic. But now we have a vote for just snippet, so I'll change it everywhere.

File: userspace/engine/rule_loader.cpp
Comment (mstemm) [reply userspace/engine/rule_loader.cpp:145]: Right, the other locations are used for the "In xxx (filename:line:col)" messages, but only the last location has a snippet.

I think as we use these messages we might find that it's helpful to have snippets from the other locations, but I think this is a good improvement (and is consistent) for now.

File: userspace/engine/rule_reader.cpp
Comment (mstemm) [reply userspace/engine/rule_reader.cpp:22]: For around 70% of the calls to decode_val, optional is false, and I was trying to keep things clean by not specifying a value for the more common case. I'll add a separate function decode_optional_val() instead. I also made the other decode_xxx functions a bit more specific to what they extracted (e.g. decode_seq is now a helper w/ decode_tags/decode_items being the direct functions, decode_exception_info_entry is a helper with decode_exception_{fields,comps,values} being the direct functions. This sets optional in the right places for all these methods.

File: userspace/engine/rule_reader.cpp
Comment (mstemm) [reply userspace/engine/rule_reader.cpp:321]: I got stuck on this too, but I believe this logic is correct. I believe the goal here is that if a rule does not have any of condition/output/desc/priority, then it *must* have enabled, and the purpose is to simply set enabled for a previously defined rule with the same name.

The old code did this by setting has_defs to true only if all of condition/output/desc/priority were defined, and if !has_defs was true, requiring that enabled was present. I think that's a bit looser than the goal--if a rule defined some but not all of condition/output/desc/priority, has_defs would be false, !has_defs would be true, and enabled would be required. I don't think that's what we want, right?

The new code only checks enabled if *none* of condition/output/desc/priority are present. Otherwise, you jump past the if and try to extract each of condition/output/desc/priority individually (and throw exceptions if any are missing/can't be decoded).

Does this sound right to you too?

File: userspace/engine/rule_reader.cpp
Comment (mstemm) [reply userspace/engine/rule_reader.cpp:297]: Sure, that's fine. I initially thought that rule_reader simply parsed the objects from yaml into a list, with a separate pass for handling append. But since this version handles append while parsing the yaml, we can keep things simpler and have append just be a local variable.

File: userspace/engine/rule_reader.cpp
Comment (mstemm) [reply userspace/engine/rule_reader.cpp:404]: Oh yes, although it should still iterate over each doc and return false if any have an error.

File: userspace/engine/falco_load_result.h
Comment (mstemm) [reply userspace/engine/falco_load_result.h:78]: Yeah, it's just verbosity. I think I'll still use separate functions for the actual implementation, but we can keep the "public" interface simple and just have a single as_string(). I think I'll actually use as_string(bool verbose) as "verbose" lines up with the command line option.

File: test/falco_tests.yaml
Comment (mstemm) [reply test/falco_tests.yaml:25]: Oh, ok! I will update the comment.

File: test/falco_tests.yaml
Comment (mstemm) [reply test/falco_tests.yaml:465]: Although I didn't want to tackle it in this PR, which is already quite big, I was hoping that someday rules loading would not always fail at the first error. The exception-based method is a good way to stop loading and unwind to a top-level location, but it will only show the first problem, not all of the potential problems in a set of rules.

With this version, the result contains any warnings up until the first error, or all warnings if there are no errors.

I think having the validate_warnings/validate_errors properties as a list will support a future change to return multiple errors/warnings if we decide to change the implementation. We'd just have to change the expected values for any tests that have multiple errors, or additional warnings after any errors.

Does that sound ok?

File: test/falco_tests.yaml
Comment (mstemm) [reply test/falco_tests.yaml:1028]: These test cases were actually duplicates--{list,rule}_append_failure were duplicates of invalid_{list,rule}_append_dangling, just without the -V argument. macro_append_failure was a duplicate of invalid_append_macro_dangling, just without the -V argument and using a different rules file.

The point of the tests was really to check the validation, so I kept the validation versions.

File: test/falco_tests.yaml
Comment (mstemm) [reply test/falco_tests.yaml:34]: Yeah, they all ultimately result in the same (single) warning LOAD_NO_EVTTYPE, but have different ways to write a condition that ultimately matches too many event types e.g. "evt.type!=execve", "not evt.type=execve", etc. There's a lot of overlap between these cases and the ones in test_filter_evttype_resolver.cpp. These tests are much older, before we even had unit tests, but I think they're still valuable to test the warning result paths.

File: test/falco_test.py
Comment (mstemm) [reply test/falco_test.py:389]: message_contains is only used for the test engine_version_mismatch, where the error message will be dependent on the current embedded falco engine version e.g. "Rules require engine version 9999999, but engine version is 14". If it's a big deal I could switch everything to be a substring match, but I thought the distinction was useful.

I believe all of the other error messages are static--they used to contain "in rule xxx" snippets but all of that info is in the context instead. The engine version is a strange dynamic thing that doesn't really fit into a context.

File: test/falco_tests.yaml
Comment (jasondellaluce) [reply test/falco_tests.yaml:460]: I see your point. I agree that `validate_warnings` and `validate_errors` must be lists to support multiple warnings and errors in the future, however I think changing our "fail at first error" model might be too much for a single PR. I'd suggest keeping the failure condition as-is for now, and eventually change it in a separate PR in which we document the rationale. Note that this would probably introduce user-facing breaking changes, so we might be to be careful about it!

File: userspace/engine/rule_loader.cpp
Comment (jasondellaluce) [reply userspace/engine/rule_loader.cpp:147]: Agree, thanks for the explanation!

File: userspace/engine/rule_reader.cpp
Comment (jasondellaluce) [reply userspace/engine/rule_reader.cpp:459]: I think this is related to failing at the first error in a ruleset file. However, I see your point here: different docs in the same file effectively represent different rulesets.

My take here is that I would return false here, and leave a todo. Then we can tackle this problem in another PR with a specific discussion. I think doing this kind of reasoning is too much for this contribution.

**Review by jasondellaluce — COMMENTED** [review]
I agree on having a `falco::` namespace. This is also a first step towards uniformity because we have `falco::application` too. Would love to have `falco::engine` one day.

I advocate on maintaining the policy of failing at the first error in a ruleset for now. It is great to support having multiple errors in the test suite and in some parts of the code, but I think this PR already brings a big contribution on the table and this problem is out of context here. The issue of reporting more than one error for each ruleset doc is a broader discussion that would involve user-facing breaking changes. It's a great addition that we should keep in our roadmap, but my advice is to open a second PR just for that and have a discussion there.

File: userspace/engine/rule_reader.cpp
Comment (mstemm) [reply userspace/engine/rule_reader.cpp:459]: I do return false now, although it does iterate over the docs and returns false if any have an error. Do you mean this should return false immediately, and skip any remaining docs in the file?

The technique of multiple docs in a file isn't really that common--none of our rules files use it. So it's not like that distinction will be that noticeable.

File: test/falco_tests.yaml
Comment (mstemm) [reply test/falco_tests.yaml:460]: Yeah I agree on not changing the rules loading code, it will not strive to return all errors. But I think it's fine to keep the test code the same, right? That's only visible internally and end users won't see the test code.

**Comment by mstemm** [comment]
> I agree on having a `falco::` namespace. This is also a first step towards uniformity because we have `falco::application` too. Would love to have `falco::engine` one day.
> 
> I advocate on maintaining the policy of failing at the first error in a ruleset for now. It is great to support having multiple errors in the test suite and in some parts of the code, but I think this PR already brings a big contribution on the table and this problem is out of context here. The issue of reporting more than one error for each ruleset doc is a broader discussion that would involve user-facing breaking changes. It's a great addition that we should keep in our roadmap, but my advice is to open a second PR just for that and have a discussion there.

Yep, I added a falco:: namespace to load_result. Further restructuring of namespaces can wait for a later PR.

And yes, returning multiple errors can also wait for a later PR. The current code will return multiple errors for the uncommon case of multiple docs in a single yaml file, unless you prefer I change it.

File: userspace/engine/rule_reader.cpp
Comment (jasondellaluce) [reply userspace/engine/rule_reader.cpp:457]: Thanks!

**Review by leogr — APPROVED** [review]
Hey @mstemm 

Thank for this PR. It's a significant improvement. 

Overall, it looks good to me. However, I've got a concern.
I don't see any real value in having non-verbose messages. The old behavior when loading rules was always to print verbose messages. It was advantageous. Think users that just copy/paste their Pod logs in a GitHub issue. Forcing them to redeploy Falco with `-v` would waste time.

That being said, I approve this PR anyway because I think it's valuable. Yet, I'd invite all @falcosecurity/falco-maintainers to reconsider verbosity mode when loading/validating rules. I believe verbosity should be enabled by default in those cases, or the previous behavior should be kept.

/approve

**Comment by mstemm** [comment]
> Hey @mstemm
> 
> Thank for this PR. It's a significant improvement.
> 
> Overall, it looks good to me. However, I've got a concern. I don't see any real value in having non-verbose messages. The old behavior when loading rules was always to print verbose messages. It was advantageous. Think users that just copy/paste their Pod logs in a GitHub issue. Forcing them to redeploy Falco with `-v` would waste time.
> 
> That being said, I approve this PR anyway because I think it's valuable. Yet, I'd invite all @falcosecurity/falco-maintainers to reconsider verbosity mode when loading/validating rules. I believe verbosity should be enabled by default in those cases, or the previous behavior should be kept.
> 
> /approve

You're right about verbose--I really thought that the old behavior was to only print full details on the error when verbose was true. But I double-checked and 0.32.1 does print details both for `-r` and -`V`. So that's a regression and I'll change that in a follow-on (small) PR.

**Comment by leogr** [comment]
> You're right about verbose--I really thought that the old behavior was to only print full details on the error when verbose was true. But I double-checked and 0.32.1 does print details both for `-r` and -`V`. So that's a regression and I'll change that in a follow-on (small) PR.

Thank you, Mark! :pray:

**Pull request merged**
